remote-storage-managers/s3/src/main/java/org/apache/kafka/rsm/s3/S3RemoteStorageManager.java

```diff
@@ -0,0 +1,659 @@
+ * is easier this way.
+ *
+ * <p>Each file is uniquely identified by three values:
+ * <ul>
+ *     <li>the base offset of the segment;</li>
+ *     <li>the last offset in the segment;</li>
+ *     <li>the epoch of the leader which uploaded the file.</li>
+ * </ul>
+ *
+ * <p>The remote log index file stores {@link RemoteLogIndexEntry}s.
+ *
+ * <p>S3 doesn't provide an ordered index for objects by an associated timestamp.
+ * However, this is needed to find the oldest segments by timestamp for
+ * {@link S3RemoteStorageManager#cleanupLogUntil(TopicPartition, long)} operation.
+ * The last modified reverse index is a timestamp index created by the storage manager itself.
+ * Each key is prefixed by the timestamp when the segment file was last modified before uploading
+ * ({@link LogSegment#lastModified()}), the rest is the usual key suffix for a segment's files.
+ * The files themselves are empty.
+ *
+ * <p>When the cleanup operation is executed, the last modified reverse index entries' list is requested from S3.
+ * Since S3 lists files in the lexicographical order, the implementation reads key names until
+ * {@code {last-modified-ms}} component is less or equal to the specified timestamp.
+ * This effectively gives the list of the segments that were modified up until the specified timestamp.
+ *
+ * <p>To keep the whole segment's file set visible for cleaning even in case of partial uploading
+ * (not all files are uploaded), last modified reverse index entries uploading must be confirmed before other files
+ * start being uploaded. The same happen on deletion: only when other files' deletion is confirmed,
+ * the last modified reverse index file is deleted.
+ *
+ * <p>The presence of a log file itself serves also as an indication that a segment has been completely uploaded and
+ * is available for use. Because of this, they are uploaded strictly after all other files and deleted strictly
+ * before all other files (i.e. awaiting S3's confirmations to keep the order).
+ * Without the log file, the whole segment is visible only for cleaning up.
+ *
+ * <p>However, there's a <strong>caveat</strong> regarding operation ordering due to S3's eventual consistency.
+ * Even if two operations happen in some strict order from the client point of view, it isn't guaranteed they will be
+ * replicated inside S3 in the same order and other clients (even the same one) will see them in the same order.
+ * In this implementation, there are no operations in which this can't be overcome by some reasonable retry policy.
+ * TODO: double check and confirm this.
+ *
+ * <p>Sometimes two brokers can think of themselves to be the leader of a partition.
+ * It's also not impossible for them to simultaneously try to upload files from a log segment.
+ * To break this tie, each file uploaded to the remote tier is suffixed with the leader epoch number.
+ * During the read time, the file set with the lower leader epoch number will be used.
```

Comment:
@satishd is there any particular reason to pick the highest over the lowest? The lowest feel safer to me since it never changes (the highest can grow).

Reply:
@ivanyu A higher leader epoch broker is safer when that common (sub)segment was uploaded with different leader epochs. It is always good to return records with higher leader epochs to avoid any scenarios involving zombie broker(which should have lower leader epoch as some other broker has become leader) thinking itself as the leader.

Reply:
@satishd got it, changed to use the highest leader epoch.